activerecord/lib/active_record/associations/has_many_through_association.rb

```diff
@@ -113,7 +113,7 @@ def remove_records(existing_records, records, method)
         end
 
         def target_reflection_has_associated_record?
-          !(through_reflection.belongs_to? && owner[through_reflection.foreign_key].blank?)
+          !(through_reflection.belongs_to? && Array(through_reflection.foreign_key).all? { |foreign_key_column| owner[foreign_key_column].blank? })
```

Comment:
We should be using a `_read_attribute` here for cases where a composite foreign key contains `id` column as `owner["id"]` returns a whole identifier and not just value of the `id` column
```suggestion
          !(through_reflection.belongs_to? && Array(through_reflection.foreign_key).all? { |foreign_key_column| owner._read_attribute(foreign_key_column).blank? })
```

Reply:
Makes sense. I'll see if I can add a regression test for this.

Reply:
Thinking about it a little more, I can't see a reason for the foreign key to ever be and id. That is what primary keys are named. I suppose we can support this use-case for completeness, but testing will be easier once I change `Cpk::Book` to have an id instead of number.

Reply:
Actually you are right, having `:id` as part of the foreign key would be a very unconventional scenario, can't say I can come up with a meaningful use-case. Let's not rush with switching to `_read_attribute` then until we have a meaningful reproduction.

Reply:
@eileencodes Nikita and I chatted about this and we agree it isn't necessary anymore to use `_read_attribute`, so I'm going to merge as-is. I'm happy to revert/fix forward in another patch if you feel strongly this use-case should be covered, but the foreign key columns should never use `id`, and I don't want to further change behaviour of associations if we aren't proving value.